mewarpx/mewarpx/emission.py

```diff
@@ -419,28 +455,87 @@ def inject_particles(self):
 
+        extra_pids = {}
+        extra_pids['E_total'] = particles_dict['E_total']
+        extra_pids['w'] = particles_dict['w']
+        if self.use_Schottky:
+            # Determine the local surface normal for each particle
+            normal_vectors = self.emitter.get_normals(
+                particles_dict['x'], particles_dict['y'], particles_dict['z']
+            )
+            extra_pids['norm_x'] = normal_vectors[:, 0]
+            extra_pids['norm_y'] = normal_vectors[:, 1]
+            extra_pids['norm_z'] = normal_vectors[:, 2]
+
         # Note some parts of WarpX call the variables ux and some parts vx,
         # and they're referred to as momenta. But I don't see anywhere
         # they're actually used as momenta including the particle mass -
         # the actual update is in Source/Particles/Pusher/UpdatePosition.H
         _libwarpx.add_particles(
-            self.species.name,
+            self.injection_species.name,
             x=particles_dict['x'],
             y=particles_dict['y'],
             z=particles_dict['z'],
             ux=particles_dict['vx'],
             uy=particles_dict['vy'],
             uz=particles_dict['vz'],
-            w=particles_dict['w'],
-            E_total=particles_dict['E_total'],
-            unique_particles=self.unique_particles
+            #w=particles_dict['w']
+            #E_total=particles_dict['E_total'],
+            unique_particles=self.unique_particles,
+            **extra_pids
         )
 
+        if self.use_Schottky:
+            # Up-weight the particles by the local Schottky factor, calculated
+            # as exp[sqrt(e / 4*pi*eps0) / (kT) * sqrt(max(-E, 0))]
+            pre_fac = (
+                np.sqrt(constants.e / (4.0 * np.pi * constants.epsilon_0))
+                / (constants.kb_eV * self.emitter.T)
+            )
+            mwxrun.calc_Schottky_weight(
+                self.injection_species.name, pre_fac
+            )
+
+            # get the total injected weight and energy
+            total_weight = 0.
+            total_energy = 0.
+            npart = 0
+            V = self.emitter.getvoltage()
+            weight_arrays = _libwarpx.get_particle_arrays(
+                self.injection_species.name, 'w', 0
+            )
+            ux_arrays = _libwarpx.get_particle_arrays(
+                self.injection_species.name, 'ux', 0
+            )
+            uy_arrays = _libwarpx.get_particle_arrays(
+                self.injection_species.name, 'uy', 0
+            )
+            uz_arrays = _libwarpx.get_particle_arrays(
+                self.injection_species.name, 'uz', 0
+            )
+            for ii, w in enumerate(weight_arrays):
+                npart += len(w)
+                total_weight += np.sum(w)
+                total_energy += np.sum(
+                    w * (0.5 * constants.m_e
+                    *(ux_arrays[ii]**2 + uy_arrays[ii]**2 + uz_arrays[ii]**2)
+                    + constants.e*V)
+                )
```

Comment:
I think you can instead call `self.emitter._get_E_total(...)` which would avoid duplicated code. I think this does it? Check though. And delete the line that fetches V above.
```suggestion
                total_energy += np.sum(self.emitter._get_E_total(
                    ux_arrays[ii], uy_arrays[ii], uz_arrays[ii],
                    constants.e, constants.m_e, w
                ))
```